reconcile/utils/gql.py

```diff
@@ -245,13 +255,32 @@ def get_resource(path: str) -> dict[str, Any]:
     return get_api().get_resource(path)
 
 
-def _init_gql_client(url: str, token: Optional[str]) -> Client:
-    req_headers = None
-    if token:
-        # The token stored in vault is already in the format 'Basic ...'
-        req_headers = {"Authorization": token}
-    # Here we are explicitly using sync strategy
-    return Client(transport=RequestsHTTPTransport(url, headers=req_headers, timeout=30))
+class PersistentRequestsHTTPTransport(RequestsHTTPTransport):
```

Comment:
I had a quick look at [RequestsHTTPTransport](https://github.com/graphql-python/gql/blob/013fa6aceb066be843849e83cc72747f51c07d0e/gql/transport/requests.py), that design is already using session, just can't change `pool_maxsize` (default 10), I'm thinking to change all `@threaded` to max 10.

According to [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)

>Changed in version 3.8: Default value of max_workers is changed to min(32, os.cpu_count() + 4). This default value preserves at least 5 workers for I/O bound tasks. It utilizes at most 32 CPU cores for CPU bound tasks which release the GIL. And it avoids using very large resources implicitly on many-core machines.

We are using 4-8 cpu cores, 10 is a proper value, and it can save us a lot of effort to customize parameters in many libs.

We just need to ensure only one Client is created in a single run, `GqlApi` init should happen in main thread, already a Singleton.

To close session after run, can use `self.client.transport.close()`, maybe worth unset `global _gqlapi` after close to do a full cleanup, or refactor Singleton like [VaultClient](https://github.com/app-sre/qontract-reconcile/blob/42bf5140ea9ae61f20d35c1ebfcabf8ec80af014/reconcile/utils/vault.py#L433) to be thread safe, can add a new static method to reset `_instance` to `None`.

In this way, we don't need to have a `global_session_cache`, just use local session and singleton instance.

Reply:
Oh, just found the session will be closed after each [execute_sync](https://github.com/graphql-python/gql/blob/013fa6aceb066be843849e83cc72747f51c07d0e/gql/client.py#L228), so we still need to customize it a little bit.

Reply:
so we can handle session inside our `GqlApi`, then inject it into `PersistentRequestsHTTPTransport`

```python
class GqlApi:
    def __init__(self):
        self.session = requests.session()
    def _init_gql_client(self) -> Client:
        transport = PersistentRequestsHTTPTransport(self.session, self.url, headers=req_headers, timeout=30)
        ...
    def close(self):
        self.session.close()

class PersistentRequestsHTTPTransport(RequestsHTTPTransport):
    def __init__(
        self,
        session: requests.Session,
        url: str,
        headers: Optional[Dict[str, Any]] = None,
        cookies: Optional[Union[Dict[str, Any], RequestsCookieJar]] = None,
        auth: Optional[AuthBase] = None,
        use_json: bool = True,
        timeout: Optional[int] = None,
        verify: Union[bool, str] = True,
        retries: int = 0,
        method: str = "POST",
        **kwargs: Any,
    ):
        super().__init__(
            url,
            headers,
            cookies,
            auth,
            use_json,
            timeout,
            verify,
            retries,
            method,
            **kwargs,
        )
        self.session = session

    def connect(self) -> None:
        pass

    def close(self) -> None:
        pass
```

Reply:
Like the idea about closing the client, I will take a look.